### ISSUE TITLE
response encoding is not utf-8 and impossible to change it

### ISSUE THREAD

**Issue by mayouf** [scrapy/scrapy#2154]
Hi everybody,

Think we have a bug here.

I am trying to scrape this page (french social and environmental data) :

`scrapy shell http://www.kelquartier.com/rhone_alpes_ain_commune_divonne_les_bains_01220-c1143/revenu_moyen.html`

I try to pick up the data in this xpath:

```
In [270]: [i.strip() for i in response.xpath('//tr[@id="carteNum_12"]//td//text()').extract()]
Out[270]: [u'', u'Densit\xe9 de logements', u'', u'1 log./ha', u'', u'', u'']
```

BUT in the page the data appear we have 3 informations in this xpath:
- **"densite de logement"**     
- **"1 log./ha"**
- **"<1 log./ha"**

Then, I noticed this specific character:   **<**

YET, when I check the body text from the response, this character is here....

```
In [303]: response.body[127220:127251]
Out[303]: '<td class="td_B">1 log./ha</td>'
```

```
In [304]: response.body[127725:127757]
Out[304]: '<td class="td_B"><1 log./ha</td>'
```

When I check the encoding, it is not utf-8 but cp1225.

```
In    [260]  : response._body_declared_encoding()
Out [260]  : 'cp1252
```

And when I checked many website that I scraped in past, the response encoding was always utf-8.

When I checked on stackoverflow, I saw many suggestion like:

`unicode(response.body.decode(response.encoding)).encode('utf-8')`

`newresponse = response.replace(encoding='utf-8')`

the encoding is always cp1252:

```
In    [260]  : response._body_declared_encoding()
Out [260]  : 'cp1252

```
and my xpath still doesn't take the information I need, which is:  '<td class="td_B">**<1 log./ha**</td>'

So in summary, this is a situation where the data is present in the body but not  in its xpath. And it only comes with pages with the following specific character **"<"**.

And trying debugging the situation, I figured out the response encoding is **cp1252** and it is impossible to change it.

**Comment by redapple** [comment]
@mayouf , I believe the encoding of the page is ISO-8859-1, close enough to cp1252.

Website response headers contain `Content-Type:text/html; charset=ISO-8859-1`
and the page itself has `<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1" />
`
In scrapy shell, you can look for "plutôt" and check how it's encoded withing the raw bytes of `response.body`:

```
$ scrapy shell http://www.kelquartier.com/rhone_alpes_ain_commune_divonne_les_bains_01220-c1143/revenu_moyen.html
2016-08-01 18:53:02 [scrapy] INFO: Scrapy 1.1.0 started (bot: scrapybot)
...
2016-08-01 18:53:02 [scrapy] INFO: Spider opened
2016-08-01 18:53:03 [scrapy] DEBUG: Crawled (200) <GET http://www.kelquartier.com/rhone_alpes_ain_commune_divonne_les_bains_01220-c1143/revenu_moyen.html> (referer: None)
...
In [1]: import re

In [2]: re.search(rb'(plut.{4})', response.body).groups()
Out[2]: (b'plut\xf4t p',)

In [3]: b'plut\xf4t p'.decode('latin1')
Out[3]: 'plutôt p'

In [4]: b'plut\xf4t p'.decode('cp1252')
Out[4]: 'plutôt p'

In [5]: b'plut\xf4t p'.decode('ISO-8859-1')
Out[5]: 'plutôt p'

In [6]: b'plut\xf4t p'.decode('utf8')
---------------------------------------------------------------------------
UnicodeDecodeError                        Traceback (most recent call last)
<ipython-input-6-f8baa01db51d> in <module>()
----> 1 b'plut\xf4t p'.decode('utf8')
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xf4 in position 4: invalid continuation byte
In [7]: 
```

The raw bytes are from latin1/cp1252 encoding, not UTF8.

Back to the `<` question,
what version of scrapy and lxml/libxml2 are you using? (paste the output of `scrapy version -v)

I'm using this:
```
$ scrapy version -v
Scrapy    : 1.1.0
lxml      : 3.6.0.0
libxml2   : 2.9.3
Twisted   : 16.2.0
Python    : 3.5.1+ (default, Mar 30 2016, 22:46:26) - [GCC 5.3.1 20160330]
pyOpenSSL : 16.0.0 (OpenSSL 1.0.2g-fips  1 Mar 2016)
Platform  : Linux-4.4.0-31-generic-x86_64-with-Ubuntu-16.04-xenial
```

and I am able to get "Densité de logements" line:

```
In [7]: for cell in response.xpath(u'//tr[normalize-space(td[1])="Densité de logements"]//td'):
    print(cell.xpath('normalize-space()').extract())
   ....:     
['Densité de logements']
['1 log./ha']
['']
['<1 log./ha']
```

It could be that the libxml version you have on your system is older than mine and having trouble parsing `<`, which should not appear like that in HTML text, rather as `&lt;` (so in theory, there's a bug in the website HTML).
The page has it correct for `Accueil > Divonne-les-Bains > Habitants - Revenu moyen` though:

```
		  <div class="fil_ariane">
            	<a href="/">Accueil</a>
                &gt;<a href="/rhone_alpes_ain_commune_divonne_les_bains_01220-c1143/revenu_moyen.html">
								Divonne-les-Bains 				</a>
                &gt; <a href="/rhone_alpes_ain_commune_divonne_les_bains_01220-c1143/revenu_moyen.html">Habitants - Revenu moyen</a>
            </div>
```

**Comment by mayouf** [comment]
here is my scrapy version:

```
mic@mic-ordi:~$ scrapy version --verbose
Scrapy    : 1.1.0
lxml      : 3.6.0.0
libxml2   : 2.9.2
Twisted   : 16.2.0
Python    : 2.7.11 |Anaconda custom (64-bit)| (default, Jun 15 2016, 15:21:30) - [GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]
pyOpenSSL : 0.15.1 (OpenSSL 1.0.2h  3 May 2016)
Platform  : Linux-4.2.0-42-generic-x86_64-with-debian-jessie-sid
```

I even did a new intstall of libxml2 and it gives me the 2.9.2 version, I can't get the same as you.

```
mic@mic-ordi:~$ conda install libxml2
Fetching package metadata .......
Solving package specifications: ..........

# All requested packages already installed.
# packages in environment at /home/mic/anaconda2:
#
libxml2                   2.9.2                         0 
```

BUT I saw something in your response: You are using python 3.5 and I do with the 2.7 one.

So here is what I did and unfortunately, it did not show a difference:

1. I created a python 3.5 environnement in Anaconda (http://conda.pydata.org/docs/py2or3.html)

`conda create -n py35 python=3.5 anaconda
`

2. I installed scrapy with the conda command:
`conda install scrapy`

3. I checked the scrapy version as you asked above:

```
(py35) mic@mic-ordi:~$ scrapy version -v
Scrapy    : 1.1.1
lxml      : 3.6.0.0
libxml2   : 2.9.2
Twisted   : 16.3.0
Python    : 3.5.2 |Anaconda 4.1.1 (64-bit)| (default, Jul  2 2016, 17:53:06) - [GCC 4.4.7 20120313 (Red 
Hat 4.4.7-1)]
pyOpenSSL : 16.0.0 (OpenSSL 1.0.2h  3 May 2016)
Platform  : Linux-4.2.0-42-generic-x86_64-with-debian-jessie-sid
```

4. I did the same command when I wrote this issue:

`scrapy shell http://www.kelquartier.com/rhone_alpes_ain_commune_divonne_les_bains_01220-c1143/revenu_moyen.html`

```
In [11]: [i.strip() for i in response.xpath('//tr[@id="carteNum_12"]//td//text()').extract()]
Out[11]: ['', 'Densité de logements', '', '1 log./ha', '', '', '']
```
```
In [12]: for cell in response.xpath(u'//tr[normalize-space(td[1])="Densité de logements"]//td'):
   ....:         print(cell.xpath('normalize-space()').extract())
   ....:     
['Densité de logements']
['1 log./ha']
['']
['']
```


5. I also check if my ipython was on the right interpreter version:

```
In [9]: import sys
In [10]: sys.version_info[0:30]
Out[10]: (3, 5, 2, 'final', 0)
```


I don't really see what is wrong here....only the libxml2 library version 2.9.2 instead of 2.9.3 is left to explain this output difference we have...

How did you get the libxml 2.9.3 library please mate because neither my pip install or my conda upgrade provides me this version ?

**Comment by nyov** [comment]
I am not sure this is a bug?
Usually in HTML/XML, `<` can not occur unescaped, it should be `&laquo;` or entity-encoded, so perhaps the parser considers it an invalid start tag in the code and eats it.
Maybe @redapple has some version or workaround of `lxml` to relax the parsing there?

Perhaps there is some way to configure `lxml.html.HTMLParser` to handle such issues, or perhaps it's something with debian's packaging of libxml2.
I don't think that the difference of behavior would be in a minor version of libxml (2 to 3), but I could be wrong.

FWIW, I'm also on debian with libxml 2.9.2
```
$ scrapy version --verbose
Scrapy    : 1.1.1
lxml      : 3.6.0.0
libxml2   : 2.9.2
Twisted   : 16.2.0
Python    : 2.7.12 (default, Jun 29 2016, 08:18:26) - [GCC 5.4.0 20160609]
pyOpenSSL : 16.0.0 (OpenSSL 1.0.2h  3 May 2016)
Platform  : Linux-3.16.0-4-amd64-x86_64-with-debian-stretch-sid
```

and I get the same results:

```
In [1]: for cell in response.xpath(u'//tr[normalize-space(td[1])="Densité de logements"]//td'):
   ...:         print(cell.xpath('normalize-space()').extract())
   ...:     
[u'Densit\xe9 de logements']
[u'1 log./ha']
[u'']
[u'']
```

**Comment by redapple** [comment]
@mayouf, @nyov , I don't know what's different with my setup.

I'm pasting console logs from a fresh Python 2.7 virtualenv creation, pip install of scrapy and scrapy shell session:

```
$ mkvirtualenv scrapy-gh-2154
Using real prefix '/usr'
New python executable in /home/paul/.virtualenvs/scrapy-gh-2154/bin/python2
Also creating executable in /home/paul/.virtualenvs/scrapy-gh-2154/bin/python
Installing setuptools, pip, wheel...done.
virtualenvwrapper.user_scripts creating /home/paul/.virtualenvs/scrapy-gh-2154/bin/predeactivate
virtualenvwrapper.user_scripts creating /home/paul/.virtualenvs/scrapy-gh-2154/bin/postdeactivate
virtualenvwrapper.user_scripts creating /home/paul/.virtualenvs/scrapy-gh-2154/bin/preactivate
virtualenvwrapper.user_scripts creating /home/paul/.virtualenvs/scrapy-gh-2154/bin/postactivate
virtualenvwrapper.user_scripts creating /home/paul/.virtualenvs/scrapy-gh-2154/bin/get_env_details

(scrapy-gh-2154) paul@laptop:~$ pip install scrapy
Collecting scrapy
  Using cached Scrapy-1.1.2-py2.py3-none-any.whl
Collecting w3lib>=1.14.2 (from scrapy)
  Using cached w3lib-1.15.0-py2.py3-none-any.whl
Collecting Twisted>=10.0.0 (from scrapy)
Collecting pyOpenSSL (from scrapy)
  Using cached pyOpenSSL-16.0.0-py2.py3-none-any.whl
Collecting parsel>=0.9.3 (from scrapy)
  Using cached parsel-1.0.3-py2.py3-none-any.whl
Collecting PyDispatcher>=2.0.5 (from scrapy)
Collecting queuelib (from scrapy)
  Using cached queuelib-1.4.2-py2.py3-none-any.whl
Collecting cssselect>=0.9 (from scrapy)
  Using cached cssselect-0.9.2-py2.py3-none-any.whl
Collecting lxml (from scrapy)
  Using cached lxml-3.6.4-cp27-cp27mu-manylinux1_x86_64.whl
Collecting service-identity (from scrapy)
  Using cached service_identity-16.0.0-py2.py3-none-any.whl
Collecting six>=1.5.2 (from scrapy)
  Using cached six-1.10.0-py2.py3-none-any.whl
Collecting zope.interface>=3.6.0 (from Twisted>=10.0.0->scrapy)
Collecting cryptography>=1.3 (from pyOpenSSL->scrapy)
Collecting pyasn1 (from service-identity->scrapy)
  Using cached pyasn1-0.1.9-py2.py3-none-any.whl
Collecting attrs (from service-identity->scrapy)
  Using cached attrs-16.0.0-py2.py3-none-any.whl
Collecting pyasn1-modules (from service-identity->scrapy)
  Using cached pyasn1_modules-0.0.8-py2.py3-none-any.whl
Requirement already satisfied (use --upgrade to upgrade): setuptools in ./.virtualenvs/scrapy-gh-2154/lib/python2.7/site-packages (from zope.interface>=3.6.0->Twisted>=10.0.0->scrapy)
Collecting cffi>=1.4.1 (from cryptography>=1.3->pyOpenSSL->scrapy)
  Using cached cffi-1.7.0-cp27-cp27mu-manylinux1_x86_64.whl
Collecting ipaddress (from cryptography>=1.3->pyOpenSSL->scrapy)
  Using cached ipaddress-1.0.16-py27-none-any.whl
Collecting enum34 (from cryptography>=1.3->pyOpenSSL->scrapy)
  Using cached enum34-1.1.6-py2-none-any.whl
Collecting idna>=2.0 (from cryptography>=1.3->pyOpenSSL->scrapy)
  Using cached idna-2.1-py2.py3-none-any.whl
Collecting pycparser (from cffi>=1.4.1->cryptography>=1.3->pyOpenSSL->scrapy)
Installing collected packages: six, w3lib, zope.interface, Twisted, pycparser, cffi, ipaddress, enum34, idna, pyasn1, cryptography, pyOpenSSL, lxml, cssselect, parsel, PyDispatcher, queuelib, attrs, pyasn1-modules, service-identity, scrapy
Successfully installed PyDispatcher-2.0.5 Twisted-16.3.2 attrs-16.0.0 cffi-1.7.0 cryptography-1.4 cssselect-0.9.2 enum34-1.1.6 idna-2.1 ipaddress-1.0.16 lxml-3.6.4 parsel-1.0.3 pyOpenSSL-16.0.0 pyasn1-0.1.9 pyasn1-modules-0.0.8 pycparser-2.14 queuelib-1.4.2 scrapy-1.1.2 service-identity-16.0.0 six-1.10.0 w3lib-1.15.0 zope.interface-4.2.0

(scrapy-gh-2154) paul@laptop:~$ scrapy version -v
Scrapy    : 1.1.2
lxml      : 3.6.4.0
libxml2   : 2.9.4
Twisted   : 16.3.2
Python    : 2.7.10 (default, Oct 14 2015, 16:09:02) - [GCC 5.2.1 20151010]
pyOpenSSL : 16.0.0 (OpenSSL 1.0.2g-fips  1 Mar 2016)
Platform  : Linux-4.4.0-34-generic-x86_64-with-Ubuntu-16.04-xenial


(scrapy-gh-2154) paul@laptop:~$ pip freeze
attrs==16.0.0
cffi==1.7.0
cryptography==1.4
cssselect==0.9.2
enum34==1.1.6
idna==2.1
ipaddress==1.0.16
lxml==3.6.4
parsel==1.0.3
pyasn1==0.1.9
pyasn1-modules==0.0.8
pycparser==2.14
PyDispatcher==2.0.5
pyOpenSSL==16.0.0
queuelib==1.4.2
Scrapy==1.1.2
service-identity==16.0.0
six==1.10.0
Twisted==16.3.2
w3lib==1.15.0
zope.interface==4.2.0

```

```
(scrapy-gh-2154) paul@laptop:~$ scrapy shell http://www.kelquartier.com/rhone_alpes_ain_commune_divonne_les_bains_01220-c1143/revenu_moyen.html
2016-08-23 13:38:54 [scrapy] INFO: Scrapy 1.1.2 started (bot: scrapybot)
2016-08-23 13:38:54 [scrapy] INFO: Overridden settings: {'LOGSTATS_INTERVAL': 0, 'DUPEFILTER_CLASS': 'scrapy.dupefilters.BaseDupeFilter'}
2016-08-23 13:38:54 [scrapy] INFO: Enabled extensions:
['scrapy.extensions.telnet.TelnetConsole',
 'scrapy.extensions.corestats.CoreStats']
2016-08-23 13:38:54 [scrapy] INFO: Enabled downloader middlewares:
['scrapy.downloadermiddlewares.httpauth.HttpAuthMiddleware',
 'scrapy.downloadermiddlewares.downloadtimeout.DownloadTimeoutMiddleware',
 'scrapy.downloadermiddlewares.useragent.UserAgentMiddleware',
 'scrapy.downloadermiddlewares.retry.RetryMiddleware',
 'scrapy.downloadermiddlewares.defaultheaders.DefaultHeadersMiddleware',
 'scrapy.downloadermiddlewares.redirect.MetaRefreshMiddleware',
 'scrapy.downloadermiddlewares.httpcompression.HttpCompressionMiddleware',
 'scrapy.downloadermiddlewares.redirect.RedirectMiddleware',
 'scrapy.downloadermiddlewares.cookies.CookiesMiddleware',
 'scrapy.downloadermiddlewares.chunked.ChunkedTransferMiddleware',
 'scrapy.downloadermiddlewares.stats.DownloaderStats']
2016-08-23 13:38:54 [scrapy] INFO: Enabled spider middlewares:
['scrapy.spidermiddlewares.httperror.HttpErrorMiddleware',
 'scrapy.spidermiddlewares.offsite.OffsiteMiddleware',
 'scrapy.spidermiddlewares.referer.RefererMiddleware',
 'scrapy.spidermiddlewares.urllength.UrlLengthMiddleware',
 'scrapy.spidermiddlewares.depth.DepthMiddleware']
2016-08-23 13:38:54 [scrapy] INFO: Enabled item pipelines:
[]
2016-08-23 13:38:54 [scrapy] DEBUG: Telnet console listening on 127.0.0.1:6023
2016-08-23 13:38:55 [scrapy] INFO: Spider opened
2016-08-23 13:38:56 [scrapy] DEBUG: Crawled (200) <GET http://www.kelquartier.com/rhone_alpes_ain_commune_divonne_les_bains_01220-c1143/revenu_moyen.html> (referer: None)
[s] Available Scrapy objects:
[s]   crawler    <scrapy.crawler.Crawler object at 0x7f14a1a35cd0>
[s]   item       {}
[s]   request    <GET http://www.kelquartier.com/rhone_alpes_ain_commune_divonne_les_bains_01220-c1143/revenu_moyen.html>
[s]   response   <200 http://www.kelquartier.com/rhone_alpes_ain_commune_divonne_les_bains_01220-c1143/revenu_moyen.html>
[s]   settings   <scrapy.settings.Settings object at 0x7f14a1a35c50>
[s]   spider     <DefaultSpider 'default' at 0x7f14a13c9290>
[s] Useful shortcuts:
[s]   shelp()           Shell help (print this help)
[s]   fetch(req_or_url) Fetch request (or URL) and update local objects
[s]   view(response)    View response in a browser
>>> for cell in response.xpath(u'//tr[normalize-space(td[1])="Densité de logements"]//td'):
...     print(cell.xpath('normalize-space()').extract())
... 
[u'Densit\xe9 de logements']
[u'1 log./ha']
[u'']
[u'<1 log./ha']
>>> print(response.xpath(u'//tr[normalize-space(td[1])="Densité de logements"]').extract_first()
... )
<tr class="hoverCarte cursor_pointer" id="carteNum_12" title="Allez à la carte : Densité de logements - Divonne-les-Bains" onclick="changeMap('/rhone_alpes_ain_commune_divonne_les_bains_01220-c1143/densite_de_logements.html');return false;">
                                       
                                                                                                                     
                                    	                                    	                                    	                                                                                										
                                        <td class="td_B">
                                        <a href="/rhone_alpes_ain_commune_divonne_les_bains_01220-c1143/densite_de_logements.html" onclick="changeMap('/rhone_alpes_ain_commune_divonne_les_bains_01220-c1143/densite_de_logements.html');return false;"> Densité de logements</a>
                                        </td>
                                        <td class="td_B">1 log./ha</td>
                                        <td align="center" class="td_B" style="width:40px !important">
											                                            	                                                <img src="../../../images/fr/fleche_nbr_horizontale.png" width="13" height="14">
                                                                                                                                    </td>
                                        <td class="td_B">&lt;1 log./ha</td>
                                    </tr>
>>> response.body.find('<1 log')
127742

>>> response.body[127700:127780]
'                         <td class="td_B"><1 log./ha</td>\r\n                     '
>>> 
```

**Comment by redapple** [comment]
@mayouf , another option is to try with [`html5lib`](http://lxml.de/html5parser.html):

```
$ pip install html5lib
Collecting html5lib
Requirement already satisfied (use --upgrade to upgrade): six in ./.virtualenvs/scrapy-gh-2154/lib/python2.7/site-packages (from html5lib)
Requirement already satisfied (use --upgrade to upgrade): setuptools>=18.5 in ./.virtualenvs/scrapy-gh-2154/lib/python2.7/site-packages (from html5lib)
Collecting webencodings (from html5lib)
Installing collected packages: webencodings, html5lib
Successfully installed html5lib-0.999999999 webencodings-0.5

(scrapy-gh-2154) paul@paul-SATELLITE-R830:~$ scrapy shell http://www.kelquartier.com/rhone_alpes_ain_commune_divonne_les_bains_01220-c1143/revenu_moyen.html
2016-08-23 13:56:37 [scrapy] INFO: Scrapy 1.1.2 started (bot: scrapybot)
(...)
2016-08-23 13:56:39 [scrapy] DEBUG: Crawled (200) <GET http://www.kelquartier.com/rhone_alpes_ain_commune_divonne_les_bains_01220-c1143/revenu_moyen.html> (referer: None)
(...)
>>> from lxml.html import html5parser
>>> response.text.find('<1 log')
127742
>>> print(response.text[127742-1000:127742+1000])
                                            										
                                        <td class="td_B">
                                        <a href="/rhone_alpes_ain_commune_divonne_les_bains_01220-c1143/densite_de_logements.html" onclick="changeMap('/rhone_alpes_ain_commune_divonne_les_bains_01220-c1143/densite_de_logements.html');return false;"> Densité de logements</a>
                                        </td>
                                        <td class="td_B">1 log./ha</td>
                                        <td align="center" class="td_B"  style="width:40px !important">
											                                            	                                                <img src="../../../images/fr/fleche_nbr_horizontale.png" width="13" height="14" />
                                                                                                                                    </td>
                                        <td class="td_B"><1 log./ha</td>
                                    </tr>
                                    
                            
                                                                        						                                                                   
                                    <tr class="hoverCarte cursor_pointer" id="carteNum_13" title="Allez à la carte : Type de maison - Divonne-les-Bains"  onclick="changeMap('/rhone_alpes_ain_commune_divonne_les_bains_01220-c1143/type_de_maison.html');return false;">
                                       
                                                                                                                     
                                    	                                    	                                    	                                                                                										
                                        <td class="td_A">
                                   
>>> doc = html5parser.fromstring(response.body)
>>> doc.xpath('//x:tr[@id="carteNum_12"]', namespaces={'x': 'http://www.w3.org/1999/xhtml'})
[<Element {http://www.w3.org/1999/xhtml}tr at 0x7f948ccda7a0>]

>>> doc.xpath('//x:tr[@id="carteNum_12"]//text()', namespaces={'x': 'http://www.w3.org/1999/xhtml'})
['\n                                       \n                                                                                                                     \n                                    \t                                    \t                                    \t                                                                                \t\t\t\t\t\t\t\t\t\t\n                                        ', '\n                                        ', u' Densit\xe9 de logements', '\n                                        ', '\n                                        ', '1 log./ha', '\n                                        ', '\n\t\t\t\t\t\t\t\t\t\t\t                                            \t                                                ', '\n                                                                                                                                    ', '\n                                        ', '<1 log./ha', '\n                                    ']

>>> doc.xpath(u'''
...     //x:tr[normalize-space(x:td[1])="Densité de logements"]
...         //x:td''', namespaces={'x': 'http://www.w3.org/1999/xhtml'})
[<Element {http://www.w3.org/1999/xhtml}td at 0x7f948ccda7e8>, <Element {http://www.w3.org/1999/xhtml}td at 0x7f948ccdac20>, <Element {http://www.w3.org/1999/xhtml}td at 0x7f948ccdae18>, <Element {http://www.w3.org/1999/xhtml}td at 0x7f948ccde098>]
>>> for cell in doc.xpath(u'''
...     //x:tr[normalize-space(x:td[1])="Densité de logements"]
...         //x:td''', namespaces={'x': 'http://www.w3.org/1999/xhtml'}):
...     print(cell.xpath('normalize-space()'))
... 
Densité de logements
1 log./ha

<1 log./ha
>>> 
```

**Comment by nyov** [comment]
@redapple, the virtualenv version works as you posted it. So I would assume at this point that debian's libxml2 version is responsible.

**Comment by redapple** [comment]
@mayouf , did you manage to get it to work in the end?
Did you try the html5lib way?

**Comment by redapple** [comment]
hi @mayouf , do you have any updates?

**Comment by nyov** [comment]
Just close it already :p
They can always reopen when it suits them.

**Comment by redapple** [comment]
Correct.

**Comment by Ziggy-drudge** [comment]
Glance this way for the answer



https://lnkr.fun/de89f8c
